tawazi/node/node.py

```diff
@@ -77,71 +78,89 @@ class ExecNode:
-        self.debug = debug
-        self.tag = tag
-        self.setup = setup
-        self.unpack_to = unpack_to
-        self.active = True
-        self.resource = resource
-
-        self.args: List[UsageExecNode] = args or []
-        self.kwargs: Dict[Identifier, UsageExecNode] = kwargs or {}
-
-        # 3. Assign the name
-        # This can be used in the future but is not particularly useful at the moment
-        self.__name__ = self.exec_function.__name__ if not isinstance(id_, str) else id_
-
-        # 4. Assign a default NoVal to the result of the execution of this ExecNode,
-        #  when this ExecNode will be executed, self.result will be overridden
-        # It would be amazing if we can remove self.result and make ExecNode immutable
-        self.result: Union[NoValType, Any] = NoVal
-        """Internal attribute to store the result of the execution of this ExecNode (Might change!)."""
-        # even though setting result to NoVal is not necessary... it clarifies debugging
+    Raises:
+        ValueError: if setup and debug are both True.
+    """
 
-        self.profile = Profile(cfg.TAWAZI_PROFILE_ALL_NODES)
+    # 1. assign attributes
+    id_: Identifier
+    exec_function: Callable[..., Any] = field(default_factory=lambda: lambda *args, **kwargs: None)
+    priority: int = 0
+    is_sequential: bool = cfg.TAWAZI_IS_SEQUENTIAL
+    debug: bool = False
+    tag: Optional[TagOrTags] = None
+    setup: bool = False
+    unpack_to: Optional[int] = None
+    resource: Resource = cfg.TAWAZI_DEFAULT_RESOURCE
+
+    args: List["UsageExecNode"] = field(default_factory=list)  # args or []
+    kwargs: Dict[Identifier, "UsageExecNode"] = field(default_factory=dict)  # kwargs or {}
+
+    # TODO: fix _active behavior!
+    _active: Union[bool, "UsageExecNode"] = field(init=False, default=True)
+
+    # 4. Assign a default NoVal to the result of the execution of this ExecNode,
+    #  when this ExecNode will be executed, self.result will be overridden
+    # It would be amazing if we can remove self.result and make ExecNode immutable
+    result: Union[NoValType, Any] = field(init=False, default=NoVal)
+    """Internal attribute to store the result of the execution of this ExecNode (Might change!)."""
+    # even though setting result to NoVal is not necessary... it clarifies debugging
+
+    profile: Profile = field(default_factory=lambda: Profile(cfg.TAWAZI_PROFILE_ALL_NODES))
+
+    def __post_init__(self) -> None:
+        """Post init to validate attributes."""
+        # type verifications
+        is_none = self.tag is None
+        is_tag = isinstance(self.tag, Tag)
+        is_tuple_tag = isinstance(self.tag, tuple) and all(isinstance(v, Tag) for v in self.tag)
+        if not (is_none or is_tag or is_tuple_tag):
+            raise TypeError(
+                f"tag should be of type {TagOrTags} but {self.tag} of type {type(self.tag)} is provided"
+            )
+
+        if not isinstance(self.priority, int):
```

Comment:
hmm, since we do a ton of validations, don't you think pydantic could be a good candidate there ?

Reply:
we tried it before. It didn't work